File: service/src/test/java/org/eclipse/keti/acs/service/policy/evaluation/ConditionEvaluationTest.java
/*******************************************************************************
 * Copyright 2018 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 *******************************************************************************/

package org.eclipse.keti.acs.service.policy.evaluation;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.apache.commons.lang.StringUtils;
import org.eclipse.keti.acs.commons.policy.condition.groovy.GroovyConditionCache;
import org.eclipse.keti.acs.commons.policy.condition.groovy.NonCachingGroovyConditionCache;
import org.eclipse.keti.acs.model.Attribute;
import org.eclipse.keti.acs.model.Condition;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.testng.AbstractTestNGSpringContextTests;
import org.springframework.test.util.ReflectionTestUtils;
import org.testng.Assert;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.DataProvider;
import org.testng.annotations.Test;

/**
 *
 * @author acs-engineers@ge.com
 */
@ContextConfiguration(classes = { NonCachingGroovyConditionCache.class })
public class ConditionEvaluationTest extends AbstractTestNGSpringContextTests {

    private static final String ATTRIBUTE_ISSUER = "acs";
    private static final Attribute SITE_SAN_RAMON = getAttribute("site", "San Ramon");
    private static final Attribute SITE_NEW_YORK = getAttribute("site", "New York");
    private static final Attribute SITE_BOSTON = getAttribute("site", "Boston");
    private static final Attribute SITE_LA = getAttribute("site", "LA");
    private static final Attribute LOCATION_SAN_RAMON = getAttribute("location", "San Ramon");
    private static final Attribute LOCATION_NEW_YORK = getAttribute("location", "New York");
    private static final Attribute DEPARTMENT_SALES = getAttribute("department", "sales");

    private PolicyEvaluationServiceImpl evaluationService;

    @Autowired
    private GroovyConditionCache conditionCache;

    @BeforeMethod
    private void setupMethod() throws Exception {
        this.evaluationService = new PolicyEvaluationServiceImpl();
        ReflectionTestUtils.setField(this.evaluationService, "conditionCache", this.conditionCache);
    }

    private static Set<Attribute> getAttributes(final Attribute attributeOne, final Attribute attributeTwo) {
        return new HashSet<>(Arrays.asList(attributeOne, attributeTwo));
    }

    private static Attribute getAttribute(final String name, final String value) {
        return new Attribute(ATTRIBUTE_ISSUER, name, value);
    }

    @Test(dataProvider = "validConditionsWithConstants")
    public void testConditionEvaluationWithConstants(final List<Condition> conditions, final boolean expectedResult) {
        Assert.assertEquals(this.evaluationService.evaluateConditions(Collections.emptySet(), Collections.emptySet(),
                StringUtils.EMPTY, conditions, StringUtils.EMPTY), expectedResult);
    }

    @DataProvider
    private Object[][] validConditionsWithConstants() {
        return new Object[][] { { Arrays.asList(new Condition("\"a\" == \"a\"")), true },
                { Arrays.asList(new Condition("\"a\" == \"b\"")), false },
                { Arrays.asList(new Condition("\"a\" == \"a\""), new Condition("\"b\" == \"b\"")), true },
                { Arrays.asList(new Condition("\"a\" == \"a\""), new Condition("\"a\" == \"b\"")), false },
                { Arrays.asList(new Condition("\"a\" == \"b\""), new Condition("\"c\" == \"b\"")), false },
                { null, true }, { Collections.emptyList(), true } };
    }

    @Test(dataProvider = "validConditionsWithVariables")
    public void testConditionEvaluationWithVariables(final Set<Attribute> resourceAttributes,
            final Set<Attribute> subjectAttributes, final String resourceURI, final List<Condition> conditions,
            final String resourceURITemplate, final boolean expectedResult) {
        Assert.assertEquals(this.evaluationService.evaluateConditions(subjectAttributes, resourceAttributes,
                resourceURI, conditions, resourceURITemplate), expectedResult);
    }

    @DataProvider
    private Object[][] validConditionsWithVariables() {
        return new Object[][] {

                { getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK), getAttributes(SITE_SAN_RAMON, SITE_BOSTON),
                        StringUtils.EMPTY,
                        Arrays.asList(new Condition("match.any(subject.attributes(\"acs\", \"site\"),"
                                + " resource.attributes(\"acs\", \"site\"))")),
                        StringUtils.EMPTY, true },

                { getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK), getAttributes(SITE_LA, SITE_BOSTON), StringUtils.EMPTY,
                        Arrays.asList(new Condition("match.any(subject.attributes(\"acs\", \"site\"),"
                                + " resource.attributes(\"acs\", \"site\"))")),
                        StringUtils.EMPTY, false },

                { getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK), getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK),
                        StringUtils.EMPTY,
                        Arrays.asList(new Condition("match.any(subject.attributes(\"acs\", \"site\"),"
                                + " resource.attributes(\"acs\", \"site\"))")),
                        StringUtils.EMPTY, true },

                { getAttributes(LOCATION_SAN_RAMON, LOCATION_NEW_YORK), getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK),
                        StringUtils.EMPTY,
                        Arrays.asList(new Condition("match.any(resource.attributes(\"acs\", \"location\"),"
                                + " subject.attributes(\"acs\", \"site\"))")),
                        StringUtils.EMPTY, true },

                { getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK), Collections.emptySet(), StringUtils.EMPTY,
                        Arrays.asList(
                                new Condition("match.single(resource.attributes(\"acs\", \"site\"), \"San Ramon\")")),
                        StringUtils.EMPTY, true },

                { Collections.emptySet(), Collections.emptySet(), "http://assets.predix.io/site/Boston",
                        Arrays.asList(new Condition("resource.uriVariable(\"site_id\").equals(\"Boston\")")),
                        "site/{site_id}", true },

                { Collections.emptySet(), Collections.emptySet(), "http://assets.predix.io/site/Boston",
                        Arrays.asList(new Condition("resource.uriVariable(\"site_id\").equals(\"New York\")")),
                        "site/{site_id}", false },

                { Collections.emptySet(), getAttributes(SITE_LA, SITE_BOSTON), "http://assets.predix.io/site/Boston",
                        Arrays.asList(new Condition("match.single(subject.attributes(\"acs\", \"site\"),"
                                + " resource.uriVariable(\"site_id\"))")),
                        "site/{site_id}", true },

                { Collections.emptySet(), getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK),
                        "http://assets.predix.io/site/Boston",
                        Arrays.asList(new Condition("match.single(subject.attributes(\"acs\", \"site\"),"
                                + " resource.uriVariable(\"site_id\"))")),
                        "site/{site_id}", false },

                { Collections.emptySet(), getAttributes(SITE_SAN_RAMON, SITE_NEW_YORK), "http://assets.predix.io/site",
                        Arrays.asList(new Condition("match.single(subject.attributes(\"acs\", \"site\"),"
                                + " resource.uriVariable(\"site_id\"))")),
                        "site/{site_id}", false },

                { Collections.emptySet(), getAttributes(SITE_BOSTON, DEPARTMENT_SALES),
                        "http://assets.predix.io/site/Boston/department/sales",
                        Arrays.asList(
                                new Condition("match.single(subject.attributes(\"acs\", \"site\"),"
                                        + " resource.uriVariable(\"site_id\"))"),
                                new Condition("match.single(subject.attributes(\"acs\", \"department\"),"
                                        + " resource.uriVariable(\"department_id\"))")),
                        "site/{site_id}/department/{department_id}", true },

                { Collections.emptySet(), getAttributes(SITE_BOSTON, DEPARTMENT_SALES),
                        "http://assets.predix.io/site/Boston/department/sales",
                        Arrays.asList(new Condition("match.single(subject.attributes(\"acs\", \"site\"),"
                                + " resource.uriVariable(\"site_id\")) and"
                                + " match.single(subject.attributes(\"acs\", \"department\"),"
                                + " resource.uriVariable(\"department_id\"))")),
                        "site/{site_id}/department/{department_id}", true } };
    }

    @Test(dataProvider = "invalidConditions", expectedExceptions = PolicyEvaluationException.class)
    public void testInvalidConditionEvaluation(final List<Condition> conditions) {
        this.evaluationService.evaluateConditions(Collections.emptySet(), Collections.emptySet(), StringUtils.EMPTY,
                conditions, StringUtils.EMPTY);
    }

    @DataProvider
    private Object[][] invalidConditions() {
        return new Object[][] { { Arrays.asList(new Condition(StringUtils.EMPTY)) },
                { Arrays.asList(new Condition("'a'.equals('a')"), new Condition("System.exit(0)")) } };
    }
}

File: service/src/test/java/org/eclipse/keti/acs/service/policy/evaluation/ConditionEvaluationTest.kt
/*******************************************************************************
 * Copyright 2018 General Electric Company
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package org.eclipse.keti.acs.service.policy.evaluation

import org.eclipse.keti.acs.commons.policy.condition.groovy.GroovyConditionCache
import org.eclipse.keti.acs.commons.policy.condition.groovy.GroovyConditionShell
import org.eclipse.keti.acs.model.Attribute
import org.eclipse.keti.acs.model.Condition
import org.eclipse.keti.acs.rest.BaseResource
import org.eclipse.keti.acs.rest.BaseSubject
import org.eclipse.keti.acs.service.policy.validation.PolicySetValidator
import org.eclipse.keti.acs.service.policy.validation.PolicySetValidatorImpl
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.testng.AbstractTestNGSpringContextTests
import org.springframework.test.util.ReflectionTestUtils
import org.testng.Assert
import org.testng.annotations.BeforeClass
import org.testng.annotations.DataProvider
import org.testng.annotations.Test
import java.util.Arrays
import java.util.HashSet

/**
 *
 * @author acs-engineers@ge.com
 */
@ContextConfiguration(
    classes = [(GroovyConditionCache::class), (GroovyConditionShell::class), (PolicySetValidatorImpl::class)]
)
class ConditionEvaluationTest : AbstractTestNGSpringContextTests() {

    @Autowired
    private lateinit var policySetValidator: PolicySetValidator

    // TODO: Should no exception just return the effect if the policy matched
    val conditions: Array<Array<Any?>>
        @DataProvider(name = "conditionsProvider")
        get() = arrayOf<Array<Any?>>(
            arrayOf(Arrays.asList(Condition("\"a\" == \"a\"")), true, false),
            arrayOf(Arrays.asList(Condition("\"a\" == \"b\"")), false, false),
            arrayOf(Arrays.asList(Condition("")), false, true),
            arrayOf(emptyList<Any>(), true, true),
            arrayOf(Arrays.asList(Condition("\"a\" == \"a\""), Condition("\"b\" == \"b\"")), true, false),
            arrayOf(Arrays.asList(Condition("\"a\" == \"a\""), Condition("\"a\" == \"b\"")), false, false),
            arrayOf(Arrays.asList(Condition("\"a\" == \"b\""), Condition("\"c\" == \"b\"")), false, false),
            arrayOf(null, true, false)
        )

    val conditionsWithVariables: Array<Array<Any?>>
        @DataProvider(name = "conditionsWithVariablesProvider")
        get() = arrayOf(

            arrayOf(
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "San Ramon"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "San Ramon"),
                        Attribute("acs", "site", "Boston")
                    )
                ),
                Arrays.asList(Condition("match.any(subject.attributes(\"acs\", \"site\")," + " resource.attributes(\"acs\", \"site\"))")),
                true,
                false,
                "",
                ""
            ),
            arrayOf(
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "San Ramon"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "LA"),
                        Attribute("acs", "site", "Boston")
                    )
                ),
                Arrays.asList(Condition("match.any(subject.attributes(\"acs\", \"site\")," + " resource.attributes(\"acs\", \"site\"))")),
                false,
                false,
                "",
                ""
            ),
            arrayOf(
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "San Ramon"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "San Ramon"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                Arrays.asList(Condition("match.any(subject.attributes(\"acs\", \"site\")," + " resource.attributes(\"acs\", \"site\"))")),
                true,
                false,
                "",
                ""
            ),
            arrayOf(
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "location", "San Ramon"),
                        Attribute("acs", "location", "New York")
                    )
                ),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "San Ramon"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                Arrays.asList(Condition("match.any(resource.attributes(\"acs\", \"location\")," + " subject.attributes(\"acs\", \"site\"))")),
                true,
                false,
                "",
                ""
            ),
            arrayOf(
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "San Ramon"),
                        Attribute("acs", "site", "New York")
                    )
                ), emptySet<Any>(), Arrays.asList(
                    Condition(
                        "match.single(resource.attributes(\"acs\", \"site\"), \"San Ramon\")"
                    )
                ), true, false, "", ""
            ),
            arrayOf(
                emptySet<Any>(),
                emptySet<Any>(),
                Arrays.asList(Condition("resource.uriVariable(\"site_id\").equals(\"boston\")")),
                true,
                false,
                "http://assets.predix.io/site/boston",
                "site/{site_id}"
            ),
            arrayOf(
                emptySet<Any>(),
                emptySet<Any>(),
                Arrays.asList(Condition("resource.uriVariable(\"site_id\").equals(\"newyork\")")),
                false,
                false,
                "http://assets.predix.io/site/boston",
                "site/{site_id}"
            ),
            arrayOf(
                emptySet<Any>(),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "boston"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                Arrays.asList(Condition("match.single(subject.attributes(\"acs\", \"site\")," + " resource.uriVariable(\"site_id\"))")),
                true,
                false,
                "http://assets.predix.io/site/boston",
                "site/{site_id}"
            ),
            arrayOf(
                emptySet<Any>(),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "LA"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                Arrays.asList(Condition("match.single(subject.attributes(\"acs\", \"site\")," + " resource.uriVariable(\"site_id\"))")),
                false,
                false,
                "http://assets.predix.io/site/boston",
                "site/{site_id}"
            ),
            arrayOf(
                emptySet<Any>(),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "LA"),
                        Attribute("acs", "site", "New York")
                    )
                ),
                Arrays.asList(Condition("match.single(subject.attributes(\"acs\", \"site\")," + " resource.uriVariable(\"site_id\"))")),
                false,
                false,
                "http://assets.predix.io/site",
                "site/{site_id}"
            ),
            arrayOf(
                emptySet<Any>(),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "boston"),
                        Attribute("acs", "department", "sales")
                    )
                ),
                Arrays.asList(
                    Condition("match.single(subject.attributes(\"acs\", \"site\")," + " resource.uriVariable(\"site_id\"))"),
                    Condition("match.single(subject.attributes(\"acs\", \"department\")," + " resource.uriVariable(\"department_id\"))")
                ),
                true,
                false,
                "http://assets.predix.io/site/boston/department/sales",
                "site/{site_id}/department/{department_id}"
            ),
            arrayOf(
                emptySet<Any>(),
                HashSet(
                    Arrays.asList(
                        Attribute("acs", "site", "boston"),
                        Attribute("acs", "department", "sales")
                    )
                ),
                Arrays.asList(
                    Condition(
                        "match.single(subject.attributes(\"acs\", \"site\"),"
                        + " resource.uriVariable(\"site_id\")) "
                        + "and match.single(subject.attributes(\"acs\", \"department\"),"
                        + " resource.uriVariable(\"department_id\"))"
                    )
                ),
                true,
                false,
                "http://assets.predix.io/site/boston/department/sales",
                "site/{site_id}/department/{department_id}"
            )
        )

    @BeforeClass
    fun setup() {
        val policySetValidatorImpl = (this.policySetValidator as PolicySetValidatorImpl)
        policySetValidatorImpl.setValidAcsPolicyHttpActions("GET, POST, PUT, DELETE, PATCH, SUBSCRIBE, MESSAGE")
        policySetValidatorImpl.init()
    }

    @Test(dataProvider = "conditionsProvider")
    fun testConditionEvaluationWithConstants(
        conditions: List<Condition>?,
        expectedResult: Boolean,
        throwsException: Boolean
    ) {
        val evaluationService = PolicyEvaluationServiceImpl()
        ReflectionTestUtils.setField(evaluationService, "policySetValidator", this.policySetValidator)
        val subjectAttributes = emptySet<Attribute>()
        try {
            Assert.assertEquals(
                evaluationService.evaluateConditions(
                    subjectAttributes, HashSet(), "",
                    conditions, ""
                ), expectedResult
            )
        } catch (e: Exception) {
            if (throwsException) {
                Assert.assertTrue(e is PolicyEvaluationException)
            }
        }

    }

    @Test(dataProvider = "conditionsWithVariablesProvider")
    fun testConditionEvaluationWithVariables(
        resourceAttributes: Set<Attribute>,
        subjectAttributes: Set<Attribute>,
        conditions: List<Condition>,
        expectedResult: Boolean,
        throwsException: Boolean,
        resourceURI: String,
        resourceURITemplate: String
    ) {

        val resource = BaseResource()
        resource.attributes = resourceAttributes

        val subject = BaseSubject()
        subject.attributes = subjectAttributes

        val evaluationService = PolicyEvaluationServiceImpl()
        ReflectionTestUtils.setField(evaluationService, "policySetValidator", this.policySetValidator)

        try {
            Assert.assertEquals(
                evaluationService.evaluateConditions(
                    subjectAttributes, resourceAttributes, resourceURI,
                    conditions, resourceURITemplate
                ), expectedResult
            )

        } catch (e: Exception) {

            if (throwsException) {
                Assert.assertTrue(e is PolicyEvaluationException)
            } else {
                Assert.fail("Unexpected exception.", e)
            }

        }

    }

}
